eventsourcing/src/main/java/org/axonframework/eventsourcing/MultiStreamableMessageSource.java

```diff
@@ -0,0 +1,384 @@
+         *                sources
+         *                again repeating until the duration has been met. This ensure the highest chance of a
+         *                consumable
+         *                message being found.
+         * @param unit    the time unit for the timeout.
+         * @return true if any stream has an available message. Otherwise false.
+         *
+         * @throws InterruptedException when the thread is interrupted before the indicated time is up.
+         */
+        @Override
+        public boolean hasNextAvailable(int timeout, TimeUnit unit) throws InterruptedException {
+            long deadline = System.currentTimeMillis() + unit.toMillis(timeout);
+            int longPollTime = timeout / 10;
+
+            while (System.currentTimeMillis() < deadline) {
+                Iterator<Map.Entry<String, BlockingStream<TrackedEventMessage<?>>>> it = messageSources.entrySet()
+                                                                                                       .iterator();
+
+                while (it.hasNext()) {
+                    Map.Entry<String, BlockingStream<TrackedEventMessage<?>>> current = it.next();
+
+                    if (it.hasNext()) {
+                        if (current.getValue().hasNextAvailable()) {
+                            return true;
+                        }
+                    } else {
+                        if (current.getValue().hasNextAvailable((int) Math
+                                .min(longPollTime, deadline - System.currentTimeMillis()), unit)) {
+                            return true;
+                        }
+                    }
+                }
+            }
+
+            return false;
+        }
+
+        /**
+         * Checks each stream to see if a message is available. If more than one stream has a message it decides which
+         * message to return using the {@link #comparator}.
+         *
+         * @return selected message using {@link #comparator}.
+         *
+         * @throws InterruptedException thrown if polling is interrupted during polling a stream.
+         */
+        @Override
+        public TrackedEventMessage<?> nextAvailable() throws InterruptedException {
+            //Return peekedMessage is available.
+            if (peekedMessage != null) {
+                TrackedEventMessage next = peekedMessage;
+                peekedMessage = null;
+                return next;
+            }
+
+            HashMap<String, TrackedSourcedEventMessage> candidateMessagesToReturn = new HashMap<>();
+
+            //block until there is a message to consume
+            while (candidateMessagesToReturn.size() == 0) {
+                peekForMessages(candidateMessagesToReturn);
+            }
+
+            //get streamId of message to return after running the comparator
+            String streamIdOfMessage =
+                    candidateMessagesToReturn.entrySet().stream().min((m1, m2) -> comparator
```

Comment:
Nit: I think we can do some indenting/new-line adjustments here to make it a bit more readable.